### PR TITLE
[MRG] Flake8 diff in pull requests

### PR DESCRIPTION
**Pull request by lesteve**
Add this as a separate entry in the Travis build matrix. For now this
entry is marked as "allowed_failures".

* add flake8_diff.sh that finds the common ancestor between the branch
and upstream/master (or origin/master) and flake8 the diff with respect
to this common ancestor
* general .travis.yml refactor to follow more closely the Travis-CI docs.
* move each build step into its own script
* additional SKIP_TESTS environment variable to be able to skip the
  tests. This is useful when you only want to do the flake8 --diff part
* Ignore 'module level import not at top of file' in setup.cfg, because
  we have plenty of those in our examples.

### CODE REVIEW

File: continuous_integration/flake8_diff.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# Travis goes the git clone with a limited depth (50 at the time of
+# writing). This may not be enough to find the common ancestor with
+# $REMOTE/master so we unshallow the git checkout
+git fetch --unshallow || echo "Unshallowing the git checkout failed"
+
+# Tackle both common cases of origin and upstream as remote
+# Note: upstream has priority if it exists
+git remote -v
+git remote | grep upstream && REMOTE=upstream || REMOTE=origin
+git fetch -v $REMOTE master:remote_master
```
Comment (lesteve) [inline continuous_integration/flake8_diff.sh]: Someone with more git-fu than me may know this. I couldn't figure out a way to get the equivalent on $REMOTE/master without checking a branch locally ...
The problem comes from the fact that Travis checks out a single branch like this:
```
git clone --depth=50 --branch=test https://github.com/lesteve/nilearn.git lesteve/nilearn
```

So even doing `git fetch origin` upstream/master, origin/master is not going to show up in `git branch -r`.

**Comment by GaelVaroquaux** [comment]
@ogrisel: can you review this PR? We are somewhat thinking of it as a POC for something that could go in scikit-learn.

File: continuous_integration/flake8_diff.sh (hunk 1 above)
Comment (ogrisel) [inline continuous_integration/flake8_diff.sh]: Maybe try: `remotes/origin/master` instead?

**Comment by lesteve** [comment]
> Coverage decreased (-0.8%) to 92.767% 

Interesting glitch in coveralls given that I haven't touched a python file in this PR. Will relaunch to see whether it is just temporary.

**Comment by lesteve** [comment]
Does someone want to review this one in more details ?

The additional build is allowed to fail in Travis so there won't be any impact on build status and we can refine the flake8 checking after having learned how useful it is.

File: continuous_integration/flake8_diff.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
+
+set -e
+
+# Travis does the git clone with a limited depth (50 at the time of
+# writing). This may not be enough to find the common ancestor with
+# $REMOTE/master so we unshallow the git checkout
+git fetch --unshallow || echo "Unshallowing the git checkout failed"
```
Comment (ogrisel) [inline continuous_integration/flake8_diff.sh]: Do you have an idea of this performance impact of a full fetch on a project like joblib and on a large project like scikit-learn?

File: setup.cfg [hunk 3]
```diff
@@ -16,3 +16,6 @@ ignore-files=(plot_.*.py|conf\.py)
 
 [wheel]
 universal=1
+
+[flake8]
+ignore=E402
```
Comment (ogrisel) [inline setup.cfg]: Maybe add a comment to explain which error it is and why it needs to be ignored in the context of flake8_diff.

**Comment by ogrisel** [comment]
LGTM

File: continuous_integration/flake8_diff.sh (hunk 2 above)
Comment (lesteve) [inline continuous_integration/flake8_diff.sh]: I haven't tried yet. The thing to bear in mind is that the Travis build that does the flake8 checks doesn't run tests so I would expect the unshallowing not to matter that much in practice. Also note that if you run this script locally the unshallowing won't need to happen (at least in most cases).

Comment (lesteve) [inline continuous_integration/flake8_diff.sh]: For completeness here are some very rough timings compared for the time `git clone --depth 50` and `git fetch --unshallow`:

              | git clone depth 50 | git unshallowing |
--------------|--------------------|------------------  |
 scikit-learn | 10s                | 40s              |
 joblib       | 1-2s               | 2-3s             |

Comment (ogrisel) [inline continuous_integration/flake8_diff.sh]: > The thing to bear in mind is that the Travis build that does the flake8 checks doesn't run tests so I would expect the unshallowing not to matter that much in practice.

Alright then!

**Comment by lesteve** [comment]
I am going to merge this one and try to monitor the allowed failures in Travis to learn more about how to refine it.

**Pull request merged**